buildscripts/download/common.sh

```diff
@@ -0,0 +1,76 @@
+#!/bin/bash
+
+#    Copyright © 2021 by The qTox Project Contributors
+#
+#    This program is libre software: you can redistribute it and/or modify
+#    it under the terms of the GNU General Public License as published by
+#    the Free Software Foundation, either version 3 of the License, or
+#    (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+check_sha256()
+{
+  if uname | grep -q Darwin; then
+    HASH_BIN="gsha256sum"
+  else
+    HASH_BIN="sha256sum"
+  fi
+
+  if ! ( echo "$1  $2" | $HASH_BIN -c --quiet --status - )
+  then
+    echo "Error: sha256 of $2 doesn't match the known one."
+    echo "Expected: $1  $2"
+    echo -n "Got: "
+    sha256sum "$2"
+    exit 1
+  else
+    echo "sha256 matches the expected one: $1"
+  fi
+}
+
+download_file()
+{
+  # Curl command wrapper to download a file to the CWD
+  local URL="$1"
+  curl -L --connect-timeout 10 -O "$URL"
+}
+
+download_verify_extract_tarball()
+{
+  # Downlaoads the tarball at URL, ensures it has an sha256sum of HASH and
+  # extracts it to the CWD
+
+  local URL="$1"
+  local HASH="$2"
+
+  # Try linux style mktemp and fallback on osx style
+  TEMPDIR=$(mktemp -d -p . 2>/dev/null || mktemp -d -t qtox_download)
+
+  if [ $? -ne 0 ]; then
+    return 1
+  fi
+
+  TEMPDIR=$(cd "$TEMPDIR"; pwd -P)
+
+  pushd "$TEMPDIR" >/dev/null || exit 1
+
+  download_file "$URL"
+
+  if ! check_sha256 "$HASH" *; then
+    rm -fr "$TEMPDIR"
+    return 1
+  fi
+
+  popd >/dev/null || exit 1
+
+  tar -xf "$TEMPDIR"/* --strip-components=1
```

Comment:
We had an issue with `tar` failing in a Docker build when `bsdtar` worked. I couldn't find it in the Issues, but here is sudden6 mentioning it https://github.com/qTox/qTox/pull/4783#issuecomment-344229953. I think @anthonybilinski was havng this issue at the time, if my memory serves right?

Anyway, this is just something to keep in mind, in case someone reports `tar` failing. Hopefully this issue has been fixed by now and it won't fail.

Reply:
I think we have fairly sufficient proof from these CI runs that a wide range of distros don't have this bug. If someone encounters this bug from running on an older system I'd say it's reasonable to tell them to upgrade their tar or work around it themselves.

Reply:
Or maybe it was Diadlo who encountered this.

>I think we have fairly sufficient proof from these CI runs that a wide range of distros don't have this bug.

CI doesn't really prove anything, the issue was never reproducible on a CI to begin with, only on user's system. It was something to do with filesystem, the system's file system was a bit unusual, or their Docker was configured to use an unusual fs, e.g. btrfs instead of overlay2. Again, I couldn't find the exact post in the issue tracker, it's somewhere in there but my search queries don't turn it up, so I can only speculate based off my vague memory at this point.

>If someone encounters this bug from running on an older system I'd say it's reasonable to tell them to upgrade their tar

It's not their tar that is causing the issue, but the tar inside the Docker container, which is the same tar everyone running the container would get.

Anyway, all I'm saying is just to keep this in mind, nothing really needs to be modified. This issue happened years ago, so perhaps it got fixed already in Docker or one of its dependencies - whatever was causing that bug.